### PR TITLE
[http-client-csharp] bump MGC version and implement the credential types

### LINKED ISSUE #45918: Consume `TokenCredential`

### PR DESCRIPTION
**Pull request by ArcturusZhang**
Fixes https://github.com/Azure/azure-sdk-for-net/issues/45918

The things we used here have been verbosely tested in [this PR](https://github.com/microsoft/typespec/pull/5231) therefore this PR does not introduce more test cases for them

### CODE REVIEW

File: eng/packages/http-client-csharp/generator/TestProjects/Local/Basic-TypeSpec/src/Generated/BasicTypeSpecClient.RestClient.cs [hunk 1]
```diff
@@ -15,21 +15,12 @@ namespace BasicTypeSpec
     public partial class BasicTypeSpecClient
     {
         private static ResponseClassifier _pipelineMessageClassifier200;
-        private static ResponseClassifier _pipelineMessageClassifier201;
```
Comment (ArcturusZhang) [inline eng/packages/http-client-csharp/generator/TestProjects/Local/Basic-TypeSpec/src/Generated/BasicTypeSpecClient.RestClient.cs:18]: these changes come from https://github.com/microsoft/typespec/pull/5174

File: eng/packages/http-client-csharp/generator/TestProjects/Local/Basic-TypeSpec/src/Generated/BasicTypeSpecClient.cs [hunk 2]
```diff
@@ -110,14 +138,15 @@ public virtual async Task<Response> SayHiAsync(string headParameter, string quer
         /// <param name="headParameter"></param>
         /// <param name="queryParameter"></param>
         /// <param name="optionalQuery"></param>
+        /// <param name="cancellationToken"> The cancellation token that can be used to cancel the operation. </param>
         /// <exception cref="ArgumentNullException"> <paramref name="headParameter"/> or <paramref name="queryParameter"/> is null. </exception>
         /// <exception cref="RequestFailedException"> Service returned a non-success status code. </exception>
-        public virtual Response<Thing> SayHi(string headParameter, string queryParameter, string optionalQuery = null)
+        public virtual Response<Thing> SayHi(string headParameter, string queryParameter, string optionalQuery = null, CancellationToken cancellationToken = default)
```
Comment (ArcturusZhang) [inline eng/packages/http-client-csharp/generator/TestProjects/Local/Basic-TypeSpec/src/Generated/BasicTypeSpecClient.cs:144]: this comes from https://github.com/microsoft/typespec/pull/5337

**Review by live1206 — APPROVED** [review]

**Review by jorgerangel-msft — APPROVED** [review]

File: eng/packages/http-client-csharp/generator/TestProjects/Local/Basic-TypeSpec/src/Generated/BasicTypeSpecClient.cs [hunk 3]
```diff
@@ -55,6 +66,23 @@ public BasicTypeSpecClient(Uri endpoint, AzureKeyCredential keyCredential, Basic
...
+        public BasicTypeSpecClient(Uri endpoint, TokenCredential tokenCredential, BasicTypeSpecClientOptions options)
+        {
+            Argument.AssertNotNull(endpoint, nameof(endpoint));
+            Argument.AssertNotNull(tokenCredential, nameof(tokenCredential));
+
+            options ??= new BasicTypeSpecClientOptions();
+
+            _endpoint = endpoint;
+            _tokenCredential = tokenCredential;
+            Pipeline = HttpPipelineBuilder.Build(options, new HttpPipelinePolicy[] { new AzureKeyCredentialPolicy(_keyCredential, AuthorizationHeader) });
```
Comment (JoshLove-msft) [inline eng/packages/http-client-csharp/generator/TestProjects/Local/Basic-TypeSpec/src/Generated/BasicTypeSpecClient.cs:83]: Shouldn't this be creating a bearer token policy using the tokenCredential?

Comment (JoshLove-msft) [reply eng/packages/http-client-csharp/generator/TestProjects/Local/Basic-TypeSpec/src/Generated/BasicTypeSpecClient.cs:83]: Let's please add at least two tests - one for KeyCredential and one for TokenCredential constructor bodies.

Comment (ArcturusZhang) [reply eng/packages/http-client-csharp/generator/TestProjects/Local/Basic-TypeSpec/src/Generated/BasicTypeSpecClient.cs:83]: !!! this means the implementation in MGC is incorrect.
Let me check and add test cases

Comment (ArcturusZhang) [reply eng/packages/http-client-csharp/generator/TestProjects/Local/Basic-TypeSpec/src/Generated/BasicTypeSpecClient.cs]: This has been fixed and test cases have been added

**Review by JoshLove-msft — APPROVED** [review]

**Pull request merged**